Feature ps 500 nginx php 5 additional services

ericjenkins opened the pull request:
This adaptation of docker-compose.yml calls for a static IP. 

Last time we tried this, it failed on Mac OSX machines. However, I'm hopeful that it will work this time now that Luc has introduced a new Virtual Machine hosting platform for Docker on Mac. Also, there is a command that can be run in Mac that is supposed allow for static IP's. I added the command to the README.md file.

**Summary of changes**

- Introducing a .env file to track environment variables
- Introducing a .gitignore file
- Updated README.md with extra documentation
- Adding phpMyAdmin container
- Adding Mailhog container
- Introducing conf.d directories for PHP and Nginx
- Adding gosu package to the PHP and Nginx containers
  - This is for handling file ownership inside containers.
- Adding entrypoint.sh files to allow bash usage as www-data
  - This is also for handling file ownership inside containers
- Adding code to entrypoint.sh to allow UID:GID carryover from host
- Using a static IP
- Removed the Backend service. I don't think it is necessary.

ericjenkins committed:
PS-500 and GS-173: Various updates to Simple Docker Drupal

- Introducing a .env file to track environment variables
- Introducing a .gitignore file
- Ignoring the web directory, for various projects' repos
- Updated README.md with extra documentation
- Adding phpMyAdmin container
- Adding Mailhog container
- Introducing conf.d directories for PHP and Nginx
- Adding gosu package to the PHP and Nginx containers
- Adding entrypoint.sh files to allow bash usage as www-data
- Adding code to entrypoint.sh to allow UID:GID carryover from host
- Using a static IP

ericjenkins committed:
PS-500: Removing the gosu feature from nginx

- This wasn't working and I don't think it's as necessary, anyways.

ericjenkins committed:
PS-500: Removing the 'backend' container from docker-compose.yml

- I don't think the backend container is necessary, since it only
  serves as a data container and this task can be handled by the
  PHP container or the Nginx container.
- I moved the data hosting responsibility to the pho container.

ericjenkins committed:
PS-500 & GS-173: Adding Redis and Solr to Docker

ericjenkins committed:
PS-500 & GS-173: Tracking a settings.local.dist.php file.

ericjenkins committed:
PS-500: Adding notes about static IP usage on Mac OSX.

ericjenkins committed:
PS-500: Adding more cleanup instructions, to save space.

ericjenkins committed:
PS-500: Adding ../../shared/private/tmp directory support.

ericjenkins committed:
PS-500: Fixing nginx file permissions issues.

- Adding back an nginx Dockerfile
- Added an entrypoint.sh script for nginx
- In the Nginx container, '/var/www' will be owned by user 'nginx'
- Tracking a settings.local.dist.php file, for standing up projects
- Updated the Readme

ericjenkins committed:
PS-500: Updating the PHP memory limit to '-1' with a conf file.

Luukyb commented on README.md:
````diff
@@ -2,37 +2,108 @@
 All your Drupal files should be placed into the folder "web".
 
+```sh
+mkdir web
````
Let's have a web folder by default. Inside that folder, one php file called index.php, with <?php phpinfo(); ?>
So straight away people can check that their php is working.

ericjenkins replied:
Okay.

Luukyb commented on README.md:
````diff
@@ -2,37 +2,108 @@
+mkdir web
+```
+
+Add the project domain to your local Hosts file, and point it to
````
Could we make this optional? If yes, let's move it bellow as "Optional". Optional domain name ... etc.
I want people to be able to start quickly, then if they want more, they would check the README part "Optional".

Luukyb commented on README.md:
````diff
@@ -2,37 +2,108 @@
+
 ```sh
-$  docker-compose run --rm php ./../vendor/bin/drush site-install
+docker-compose run --rm php ./../vendor/bin/drush site-install
````
We are not using vendor anymore.

Luukyb commented on README.md:
```diff
@@ -2,37 +2,108 @@
+- User: drupal
+- Password: drupal
+- Database: drupal
+- Host: db.drupal.dev
```
Only "db" does not work anymore?

ericjenkins replied:
"db" works, but only if Mac OSX is able to support static IP addresses. This solution has been working really well for me on my local, because I can execute Drush commands both inside and outside the php container. And I can still run multiple Docker-based sites side-by-side.

If we end up needing to use random ports as we discussed at our last weekly meeting, then we would have to give up on using this "hostname"-based solution for accessing the database, and then Drush could only be executed from inside the php container.

Luukyb commented on README.md:
````diff
@@ -2,37 +2,108 @@
+docker-compose run --rm php bash
+```
+
+- Windows users: This is unavailable, for now. Docker Compose for Windows
````
In my experience, it is supported. Could we test this again?
I had student using Docker-compose on W10 pro edition. It did not work however with lower versions of W10.

ericjenkins replied:
I have not tried Interactive Mode on Windows for quite some time. Hopefully you are right! Being that interactive mode might be supported now, I will remove this note. If I test in Windows on an updated Docker server and continue to have problems, I could always re-add the note.

Luukyb commented on README.md:
````diff
@@ -2,37 +2,108 @@
+docker-compose down   # Remove containers, networks, images, and volumes.
+```
+
+#### Using Docker for system-wide management & cleanup
````
This could be a separate document.

ericjenkins replied:
Understood. I do like to refer to these commands somewhat frequently. But, if you prefer, we could start another readme file for this additional documentation. Perhaps located at `/docker/README.md` ?

Luukyb commented on docker-compose.yml:
```diff
@@ -1,23 +1,52 @@
 version: '2'
 services:
-  backend:
-    image: debian:jessie
```
Where did this go?

ericjenkins replied:
I don't think we need the 'backend' image, because we can assign data volumes to the `php` container and then simply use `volumes_from` to point other containers to php. I felt removing this would simplify the docker-compose.yml file a bit.

Luukyb replied:
@dsdeiz any input?

Luukyb commented on docker-compose.yml:
```diff
@@ -1,23 +1,52 @@
+  redis:
+    image: redis
+  solr:
+    image: solr
```
Should we use a tag?

ericjenkins replied:
I could break out a new branch for my GoinGlobal environment. That way, I could remove these from the original feature branch, and I could assign specific tags for these that apply to GoinGlobal.

ericjenkins replied:
I will remove references to Redis and Solr in this branch. I could add them back in in another branch.

Luukyb commented on README.md:
```diff
@@ -2,37 +2,108 @@
+- [Drupal Site](http://drupal.dev)
+- [phpMyAdmin](http://drupal.dev:8080)
+- [MailHog](http://drupal.dev:8025)
+- [Solr Admin](http://drupal.dev:8983)
```
I thought we would focus on PhpMyAdmin and Mailhog. No SOLR or anything else yet

ericjenkins replied:
Noted above: I could break this into two separate branches where one is for GoingGlobal and one is for phpMyAdmin & Mailhog, and nothing more.

Luukyb commented on docker-compose.yml:
```diff
@@ -1,23 +1,52 @@
+    links:
+      - db:db.${PROJECT}.dev
+      - mailhog
+      - redis
```
I thought we would focus on PhpMyAdmin and Mailhog. No SOLR or anything else yet

ericjenkins committed:
PS-500: Addressing domain related errors thrown by Drush for Solr.

ericjenkins committed:
PS-500: Removing Redis & Solr (and related documentation), for now.

ericjenkins committed:
PS-500: Restoring `/web` directory, updating documentation.

Luukyb commented on docker/nginx/Dockerfile:
```diff
@@ -1,5 +1,26 @@
 FROM nginx
 
-ADD default.conf /etc/nginx/conf.d/default.conf
+# Install & configure gosu:
```
Why do we need this?

Luukyb commented on docker/nginx/conf.d/default.conf:
```diff
@@ -1,6 +1,6 @@
 server {
     server_name localhost;
-    root /var/www/web; ## <-- Your only path reference.
+    root /var/www/html/web; ## <-- Your only path reference.
```
Why did it move?

ericjenkins replied:
I moved the docroot in by one directory level to deal with two issues I was having:

1. For some reason in Windows, Docker was refusing to mount the `/var/www` directory. I believe it was because the `/var/www` directory is not empty inside of Docker, and even though Docker for Linux doesn't care, this was a problem on Windows.
2. `/var/www` is considered the home directory for user `www`. As a result, any temporary files that are generated such as `.bash_history` and `.tmp` were showing up in my local system's project directory. Moving the mounted directory in one level to `/var/www/html` fixed this problem for me.

Luukyb commented on docker/nginx/entrypoint.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
What does this file does?

Luukyb commented on docker/php/Dockerfile:
```diff
@@ -1,32 +1,60 @@
-FROM php:5-fpm
+FROM php:5.6-fpm
```
Actually the branch is for PHP 5, not specially php 5.6.

Luukyb commented on docker/php/Dockerfile:
```diff
@@ -1,32 +1,60 @@
+    && export GOPATH=/opt/go \
+    && go get github.com/mailhog/mhsendmail
+
+# Install & configure gosu:
```
Why do we need gosu again?

ericjenkins replied:
The `php` and `nginx` packages do not allow usage of the `sudo` command. But, I need to be able to execute scripts other than `root`, so that file permissions are not overtaken by the uid:gid values of `0:0` (root:root). Using `gosu` combined with `entrypoint.sh` allows us to deal with various file permissions issues, such as directories not being writable by user `nginx` and files created inside the Docker containers being owned by the root user.

This approach was based in large part on this blog post: https://denibertovic.com/posts/handling-permissions-with-docker-volumes/

Luukyb commented on docker/php/Dockerfile:
```diff
@@ -1,32 +1,60 @@
-    && docker-php-ext-install gd mbstring opcache pdo pdo_mysql pdo_pgsql zip
-# set recommended PHP.ini settings
-# see https://secure.php.net/manual/en/opcache.installation.php
-RUN { \
```
I originally took this from the official docker image, it was required. Did this moved somewhere else?

Luukyb replied:
ok found it in docker/php/conf.d/opcache.ini

ericjenkins replied:
Correct. I chose to break out config files into directories to make them easier to find and easier for devs to add extra config settings. I did this for both PHP and Nginx.

Luukyb commented on docker/php/entrypoint.sh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+# Change user:group id values for 'www-data:www-data' to '1000:1000':
+usermod -u $(stat -c '%u' /var/www/html) www-data
```
You are changing it in the dockerfile. Maybe only one time?

Luukyb commented on docker/settings.local.dist.php:
```diff
@@ -0,0 +1,33 @@
+$conf['drupal_http_request_fails'] = FALSE;
+$conf['smtp_on'] = 0;
+$conf['smtp_password'] = '';
+$conf['googleanalytics_account'] = 'UA-11111111-1';
```
I like this 👍

ericjenkins committed:
PS-500. Addressing issues pointed out by Tere

- Adding entrypoint conditionals to address if the uid:gid of the
   project directory is set to 0:0.

ericjenkins committed:
PS-500. Breaking out sub-section of README.md to docker/README.md

- Moving some of the Readme notes to another sub-Readme for cleanliness.
